drivers/timer/apic_tsc.c

```diff
@@ -39,63 +101,103 @@ static ALWAYS_INLINE uint64_t rdtsc(void)
 	return lo + (((uint64_t)hi) << 32);
 }
 
+static inline void wrmsr(int32_t msr, uint64_t val)
+{
+	uint32_t hi = (uint32_t) (val >> 32);
+	uint32_t lo = (uint32_t) val;
+
+	__asm__ volatile("wrmsr" :: "d"(hi), "a"(lo), "c"(msr));
+}
+
+static void set_trigger(uint64_t deadline)
+{
+	if (IS_ENABLED(CONFIG_APIC_TSC_DEADLINE_TIMER)) {
+		wrmsr(IA32_TSC_DEADLINE_MSR, deadline);
+	} else {
+		/* use the timer ICR to trigger next interrupt */
+		uint64_t curr_cycle = rdtsc();
+		uint64_t delta_cycles = deadline - MIN(deadline, curr_cycle);
+		uint64_t icr = (delta_cycles * APIC_TIMER_TSC_M) / APIC_TIMER_TSC_N;
+
+		/* cap icr to 32 bits */
+		icr = MIN(icr, UINT32_MAX);
+		x86_write_loapic(LOAPIC_TIMER_ICR, icr);
+	}
+}
+
 static void isr(const void *arg)
 {
 	ARG_UNUSED(arg);
+
 	k_spinlock_key_t key = k_spin_lock(&lock);
-	uint32_t ticks = (rdtsc() - last_announce) / CYC_PER_TICK;
+	uint64_t curr_cycle = rdtsc();
+	uint64_t delta_cycles = curr_cycle - last_cycle;
+	uint32_t delta_ticks = (cycle_diff_t)delta_cycles / CYC_PER_TICK;
 
-	last_announce += ticks * CYC_PER_TICK;
-	k_spin_unlock(&lock, key);
-	sys_clock_announce(ticks);
+	last_cycle += (cycle_diff_t)delta_ticks * CYC_PER_TICK;
+	last_tick += delta_ticks;
+	last_elapsed = 0;
 
 	if (!IS_ENABLED(CONFIG_TICKLESS_KERNEL)) {
-		sys_clock_set_timeout(1, false);
-	}
-}
+		uint64_t next_cycle = last_cycle + CYC_PER_TICK;
 
-static inline void wrmsr(int32_t msr, uint64_t val)
-{
-	uint32_t hi = (uint32_t) (val >> 32);
-	uint32_t lo = (uint32_t) val;
+		set_trigger(next_cycle);
+	}
 
-	__asm__ volatile("wrmsr" :: "d"(hi), "a"(lo), "c"(msr));
+	k_spin_unlock(&lock, key);
+	sys_clock_announce(delta_ticks);
 }
 
 void sys_clock_set_timeout(int32_t ticks, bool idle)
```

Comment:
one issue captured on ISH here, there's a chance that sys_clock_set_timeout() is called with ticks=0. Then ICR is programmed with zero, so the timer stops.

Reply:
Fixed.